modules/assim.sequential/R/SDA_OBS_Assembler.R

```diff
@@ -0,0 +1,175 @@
+#' Assembler for preparing obs.mean and obs.cov for the SDA workflow
+#'
+#' @param settings_dir the path of settings.xml object.
+#' @param Var Variable name, currently support: SMP, AGB, and LAI.
+#' @param OutDir the path to store obs.mean and obs.cov
+#' @param Obs_Prep if your settings object doesn't contain Obs_Prep, you can import it separately (details see L17-18).
+#' @param skip_buffer flag to skip calculating min var based on buffer area for agb data.
+#'
+#' @return list of obs.mean and obs.cov
+#' @export
+#' @author Dongchen Zhang
+#'
+#' @examples
+#' \dontrun{
+#' settings_dir <- "/projectnb/dietzelab/dongchen/All_NEON_SDA/NEON42/SDA/pecan.xml"
+#' OutDir <- "/projectnb/dietzelab/dongchen/All_NEON_SDA/test_OBS"
+#' Var <- c("SMP", "LAI", "AGB")
+#' OBS <- SDA_OBS_Assembler(settings_dir, Var, OutDir)
+#' }
+
+
+SDA_OBS_Assembler <- function(settings_dir, Var, OutDir, Obs_Prep = NULL, skip_buffer = TRUE){
+  #export special operator
+  `%>%` <- magrittr::`%>%` 
+  `%m+%` <- as.function(lubridate::`%m+%`)
+  
+  #read settings
+  settings <- PEcAn.settings::read.settings(settings_dir)
+  
+  # I prefer to include Obs_Prep info in the settings object, that will make life easier!
+  # if you don't like it, you can provide the Obs_Prep object 
+  # followed strictly by L76 - L82 in "/pecan/modules/assim.sequential/inst/MultiSite-Exs/SDA/Create_Multi_settings.R"
+  if(!is.null(settings$state.data.assimilation$Obs_Prep)){
+    Obs_Prep <- settings$state.data.assimilation$Obs_Prep
+  }else if(is.null(Obs_Prep)){
+    PEcAn.logger::logger.info("Please provide info of Obs_Prep object!")
+    return(0)
+  }
+  
+  #prepare site_info offline, because we need to submit this to GEO, which doesn't support Bety connection.
+  
+  Site_Info <- list(site_id = settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('id') %>% unlist() %>% as.character(),
```

Comment:
This particular bit of code is complex. I'd add a comment explaining what all of this is trying to do (I'm guessing that you're pulling out a table of site info from the settings object). A few other comments on the code itself. 
1. is there a reason this is being put into a list instead of a data.frame or tibble?
2. it's not clear to me that, as written, this command couldn't get corrupted in a way where the site_id, site_name, and lat/lon  for a specific row of data don't point to the same site (e.g. if a site in the middle of the list didn't have a site name, everything after that would be off by 1). I think you could rewrite this in a way that dives into the settings -> run -> site and then extracts all 4 variables at once.
3. why is site_name being converted to numeric?

Reply:
1. the site_info list is widely used across many pre-existed pecan functions, so we must stick with that design.
2. the site_name is just a replicated vector "rep("name", length of site)", so it won't corrupt because we will definitely need the site_id, lon, and lat.
3. because it's representing the length of sites we have.